test/backend/x11/conftest.py

```diff
@@ -0,0 +1,193 @@
+        else:
+            self.xoffset = xoffset
+
+        self.proc = None  # Handle to Xephyr instance, subprocess.Popen object
+        self.display = None
+        self.display_file = None
+
+    def __enter__(self):
+        try:
+            self.start_xephyr()
+        except:  # noqa: E722
+            self.stop_xephyr()
+            raise
+
+        return self
+
+    def __exit__(self, _exc_type, _exc_val, _exc_tb):
+        self.stop_xephyr()
+
+    def start_xephyr(self):
+        """Start Xephyr instance
+
+        Starts the Xephyr instance and sets the `self.display` to the display
+        which is used to setup the instance.
+        """
+        # get a new display
+        display, self.display_file = xcffib.testing.find_display()
+        self.display = ":{}".format(display)
+
+        # build up arguments
+        args = [
+            "Xephyr",
+            "-name",
+            "qtile_test",
+            self.display,
+            "-ac",
+            "-screen",
+            "{}x{}".format(self.width, self.height),
+        ]
+        if self.two_screens:
+            args.extend(["-origin", "%s,0" % self.xoffset, "-screen",
+                         "%sx%s" % (SECOND_WIDTH, SECOND_HEIGHT)])
+        if self.xinerama:
+            args.extend(["+xinerama"])
+
+        self.proc = subprocess.Popen(args)
+
+        if can_connect_x11(self.display, ok=lambda: self.proc.poll() is None):
+            return
+
+        # we weren't able to get a display up
+        if self.proc.poll() is None:
+            raise AssertionError("Unable to connect to running Xephyr")
+        else:
+            raise AssertionError(
+                "Unable to start Xephyr, quit with return code "
+                f"{self.proc.returncode}"
+            )
+
+    def stop_xephyr(self):
+        """Stop the Xephyr instance"""
+        # Xephyr must be started first
+        if self.proc is None:
+            return
+
+        # Kill xephyr only if it is running
+        if self.proc.poll() is None:
+            # We should always be able to kill xephyr nicely
+            self.proc.terminate()
+        self.proc.wait()
+
+        self.proc = None
+
+        # clean up the lock file for the display we allocated
+        try:
+            self.display_file.close()
+            os.remove(xcffib.testing.lock_path(int(self.display[1:])))
+        except OSError:
+            pass
+
+
+@contextlib.contextmanager
+def x11_environment(**kwargs):
+    """This backend needs a Xephyr instance running"""
+    with xvfb():
+        with Xephyr(**kwargs) as x:
+            yield x
+
+
+@pytest.fixture(scope="function")
+def xmanager(request, xephyr):
+    """
+    This replicates the `manager` fixture except that the x11 backend is hard-coded. We
+    cannot simply parametrize the `backend_name` fixture module-wide because it gets
+    parametrized by `pytest_generate_tests` in test/conftest.py and only one of these
+    parametrize calls can be used.
```

Comment:
Should we refactorize and add a `tests.helpers:create_manager(backend)` function (not fixture) that is called by the different fixtures that create a manager? IIRC we already have that code replicated a few times.

Reply:
Good idea - I've significantly reduced down the code in the `manager` and co fixtures, moving the temporary file code inside `TestManager`'s own context manager, and deleting some unnecessary lines assigning attributes to the test `config` from the default config, which is redundant as this happens anyway within the `Qtile` manager's thread.